checker/decls/scopes.go

```diff
@@ -16,83 +16,75 @@ package decls
 
 import exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
 
-// Scopes represents nested Decl environments.
-// A Scopes structure is a stack of Groups (the highest array index
-// is the top of stack), with the top of the stack being the "innermost"
-// scope, and the bottom being the "outermost" scope.  Each group is a mapping
-// of names to Decls in the ident and function namespaces.
-// Lookups are performed such that bindings in inner scopes shadow those
-// in outer scopes.
+// Scopes represents nested Decl sets where the Scopes value contains a Groups containing all
+// identifiers in scope and an optional parent representing outer scopes.
+// Each Groups value is a mapping of names to Decls in the ident and function namespaces.
+// Lookups are performed such that bindings in inner scopes shadow those in outer scopes.
 type Scopes struct {
-	scopes []*Group
+	parent *Scopes
+	scopes *Group
 }
 
 // NewScopes creates a new, empty Scopes.
 // Some operations can't be safely performed until a Group is added with Push.
 func NewScopes() *Scopes {
 	return &Scopes{
-		scopes: []*Group{},
+		scopes: newGroup(),
 	}
 }
 
-// Push adds an empty Group as the new innermost scope.
-func (s *Scopes) Push() {
-	g := newGroup()
-	s.scopes = append(s.scopes, g)
+// Push creates a new Scopes value which references the current Scope as its parent.
+func (s *Scopes) Push() *Scopes {
+	return &Scopes{
+		parent: s,
+		scopes: newGroup(),
+	}
 }
 
-// Pop removes the innermost Group from Scopes.
-// Scopes should have at least one Group.
-func (s *Scopes) Pop() {
-	s.scopes = s.scopes[:len(s.scopes)-1]
+// Pop returns the parent Scopes value for the current scope, or the current scope if the parent
+// is nil.
+func (s *Scopes) Pop() *Scopes {
+	if s.parent != nil {
+		return s.parent
+	}
+	// TODO: Consider whether this should be an error / panic.
+	return s
 }
 
-// AddIdent adds the ident Decl in the outermost scope.
```

Comment:
The previous behavior here worked for testing, but could have resulted in cases like the following passing the check when they shouldn't have:

```javascript
// note, x is never declared but it's type is inferred as an int
list.exists(x, x > 10) && x != 5
```